Rework button coloring, add focus and active colors

silverwind opened the pull request:
We were missing overrides for `:focus` and `:active` styles which I've added here along with two new color variants `dark-1` and `dark-2` for them. Fomantic UI has 4 different colors but I think 3 are sufficient. I also changed it on arc-green so button goes darker when pressed.

<img width="129" alt="Screenshot 2023-05-04 at 01 21 43" src="https://user-images.githubusercontent.com/115237/236072060-7389276a-275b-4d3e-aa52-20b37c6e6d92.png">
<img width="130" alt="Screenshot 2023-05-04 at 01 17 59" src="https://user-images.githubusercontent.com/115237/236071818-0e46414a-33db-4bb2-a3bd-35b514a8a2d0.png">
<img width="129" alt="Screenshot 2023-05-04 at 01 18 07" src="https://user-images.githubusercontent.com/115237/236071819-562b1e38-541f-432b-b3b6-48e6d7594d00.png">
<img width="131" alt="Screenshot 2023-05-04 at 01 18 13" src="https://user-images.githubusercontent.com/115237/236071820-89b7dba9-ce6c-48e5-a075-9053063e6ad3.png">
<img width="133" alt="Screenshot 2023-05-04 at 01 18 30" src="https://user-images.githubusercontent.com/115237/236071823-b6fe2df4-b3f0-4dc8-97a8-f90ba6d19bec.png">
<img width="133" alt="Screenshot 2023-05-04 at 01 18 40" src="https://user-images.githubusercontent.com/115237/236071824-b02ce61a-2367-4c29-8a25-45f231f5e5ee.png">

One misc change includes some fixes to editor and slightly darker selection.

<img width="1245" alt="Screenshot 2023-05-28 at 19 16 19" src="https://github.com/go-gitea/gitea/assets/115237/1ea4a4b6-26ba-45af-9cbc-5b8c476c2338">

wxiaoguang commented:
I would prefer to reduce the colors .... if it is feasible (well, I haven't seen why it's not feasible, I think it's better to be restraint instead of bloating the code)

* https://github.com/go-gitea/gitea/issues/19353

HesterG commented:
The `:active` and `:focus` for red basic buttons are not taking effect, looks like they are overwritten by fomantic `:hover` and `:focus` styles when all these three pseudo-classes exist. For example, `:active`:

https://user-images.githubusercontent.com/17645053/236098214-2626d286-71d4-4ca6-a988-37e27957bfb9.mov

silverwind commented:
> I would prefer to reduce the colors .... if it is feasible (well, I haven't seen why it's not feasible, I think it's better to be restraint instead of bloating the code)

Not feasible here untils `color-mix`, unless we want to accept the suboptimal state of 2 colors from us and 2 colors from fomantic.

> The :active and :focus for red basic buttons are not taking effect

Thanks, will check out.

wxiaoguang commented:
Are we using `olive` or `violet` or `pink` buttons? Many colors are never used, nobody knows whether they work as expected.

I think removing unused colors is feasible, just list all supported colors on the devtest page, and test them , show them to contributors, and remove all others.

silverwind commented:
They are still in the Fomantic CSS and can not be removed there. Generally I agree that we should probably only ever use primary,green and red buttons (or better, positive, negative), but that's a bigger refactor.

silverwind commented:
> The `:active` and `:focus` for red basic buttons are not taking effect, looks like they are overwritten by fomantic `:hover` and `:focus` styles when all these three pseudo-classes exist. For example, `:active`:
> Untitled.mov

Fixed and also added rules for `positive` and `negative`, which we should eventually use instead of `red` and `green`.

silverwind commented:
> Personally I think that the palettes for the dark and light theme should be separated

I don't understand. We define color variables for everything and a theme can override everything. The `-dark` in the color names only indicates it's a darker shade of the original color. These won't be necessary any more for themes to define come `color-mix`.

> Also… is there a reason why :focus and :active are separate colours

Personally I like the button giving me visual feedback when I pressed it. `hover` and `focus` are treatet the same, so it's 3 colors, normal, hover/focus and active. Focus could arguable be moved to an outline effect to make it easier to track focus when tabbing but I will defer that to later.

silverwind commented:
Ready for review.

lunny commented:
The buttons @HesterG mentioned have still no effection.

silverwind commented:
> The buttons @HesterG mentioned have still no effection.

Odd, I had tested them specifically.

HesterG commented:
> > The buttons @HesterG mentioned have still no effection.
> 
> Odd, I had tested them specifically.

Looks like color of texts is working correctly, but background is not for active.

silverwind commented:
Ah, I was following [Fomantic UI](https://fomantic-ui.com/elements/button.html#basic) closely and they don't alter background either on basic buttons. I guess we can add that as an "opinionated" change of ours, if it's desired.

silverwind commented:
Ideally I would like to replicate [these outline buttons](https://chakra-ui.com/docs/components/button/usage#button-variants) which show a very desaturated color on background on hover/active. To avoid having to define another 20+ CSS variables, we could theoretically use `color-mix` to obtain the shade, possibly with a graceful degradation to a grey shade for browsers without support.

silverwind commented:
Actually I think I will defer adding background highlight to basic buttons to another PR. Our basic buttons never had this background, fomantic does not have it and `color-mix` experiments are ought to be rather "experimental". So let's land this now, I will do the background addition in another PR.

silverwind commented:
Moved the button styles to its own file and consolidated them while fixing a number of issues related to primary buttons which were not yet conforming the new color schemes.

Changes include:

- Primary button now follows same color scheme as other colors, which means I had to swap the dark/light meaning for dark themes. This is a breaking change to existing dark themes, but a neceessary one.
- Some fixes around code editor, remove double border
- Overwrite the font-weight: 500 on button from fomantic to 400, which imho looks better.

silverwind commented:
Will rebase after https://github.com/go-gitea/gitea/pull/24692 is in.

silverwind commented:
I think I might need to introduce specific button color variables for primary only. Reason is that buttons are the one exception where they go darker on both themes. Other color variations are generally inverse. With separate colors, it'll allow use to avoid the breaking change too, which is nice.

silverwind commented:
I will do a similar CSS variable scheme for primary button states like https://github.com/go-gitea/gitea/pull/24790 did for variables.

silverwind commented:
Ready again. The secondary button is now also styles for the first time by us. Previously it was default fomantic's black:

<img width="601" alt="Screenshot 2023-05-28 at 19 12 56" src="https://github.com/go-gitea/gitea/assets/115237/9d61fcfb-5948-407c-a986-c15c537372fe">
<img width="601" alt="Screenshot 2023-05-28 at 19 11 19" src="https://github.com/go-gitea/gitea/assets/115237/cbe10b48-6603-49eb-8424-d3975455764c">

silverwind commented:
> Are we using olive or violet or pink buttons? Many colors are never used, nobody knows whether they work as expected.

BTW, `teal` buttons are in used a few times, but I agree we should eventually remove these colors for a more consistent UI. Maybe will need to patch fomantic to rip out those color styles.

wxiaoguang commented on templates/repo/clone_script.tmpl:
```diff
@@ -14,8 +14,12 @@
+			httpsBtn.classList[!isSSH ? 'remove' : 'add']('basic');
+		}
+		if (sshBtn) {
+			sshBtn.classList[isSSH ? 'add' : 'remove']('primary');
+			sshBtn.classList[isSSH ? 'remove' : 'add']('basic');
```
toggle?

silverwind replied:
Can do, but no strong preference because if those buttons ever get into inconsistent state, these explicit remove/add would be more robust.

wxiaoguang replied:
>  because if those buttons ever get into inconsistent state, these explicit remove/add would be more robust.

Sorry I didn't the point.  IMO the `toggle` is robust enough and clearer.

```js
sshBtn.classList.toggle('primary', isSSH)
sshBtn.classList.toggle('basic', !isSSH)
```

silverwind replied:
The point is if the first statement executes, but the second doesn't, the button will be in inconsistent state. But I guess practically this will never happen, so we can do the toggle for shorter and more readable code.

wxiaoguang replied:
> The point is if the first `classList` statement executes, but the second doesn't

Why the second doesn't execute?

https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle

`toggle(token, force): force: If included, turns the toggle into a one way-only operation. If set to false, then token will only be removed, but not added. If set to true, then token will only be added, but not removed.`

silverwind replied:
Ah yes, with that force flag it will behave the same.
